rust/tableau_summary/src/twb/raw/datasource.rs

```diff
@@ -0,0 +1,140 @@
+pub mod object_graph;
+pub mod columns;
+pub mod dep;
+pub mod substituter;
+
+#[derive(Serialize, Deserialize, Default, PartialEq, Clone, Debug)]
+pub struct RawDatasource {
+    pub name: String,
+    pub version: String,
+    pub caption: String,
+    pub connection: Option<Connection>,
+    pub column_set: ColumnSet,
+    pub object_graph: ObjectGraph,
+    pub dependencies: HashMap<String, Dep>,
+}
+
+impl<'a, 'b> From<Node<'a, 'b>> for RawDatasource {
+    fn from(n: Node) -> Self {
+        if n.get_tag() != "datasource" {
+            info!("trying to convert a ({}) to datasource", n.get_tag());
+            return Self::default();
+        }
+        Self {
+            name: n.get_maybe_attr(NAME_KEY)
+                .unwrap_or_else(|| n.get_attr("formatted-name")),
+            version: n.get_attr(VERSION_KEY),
+            caption: n.get_attr(CAPTION_KEY),
+            connection: n.get_tagged_child("connection")
+                .map(Connection::from),
+            column_set: get_column_set(n),
+            object_graph: n.get_tagged_child("_.fcp.ObjectModelEncapsulateLegacy.true...object-graph")
+                .map(ObjectGraph::from)
+                .unwrap_or_default(),
+            dependencies: n.find_tagged_children("datasource-dependencies")
+                .into_iter()
+                .map(Dep::from)
+                .map(|d| (d.name.clone(), d))
+                .collect(),
+        }
+    }
+}
+
+pub(crate) fn parse_datasources(datasources_node: Node) -> anyhow::Result<Vec<RawDatasource>> {
+    let mut datasources = datasources_node.find_all_tagged_decendants("datasource")
+        .into_iter()
+        .map(RawDatasource::from)
+        .collect::<Vec<_>>();
+    let captions = datasources.iter()
+        .map(|d| (d.name.clone(), d.caption.clone()))
+        .collect::<HashMap<_, _>>();
+    datasources.iter_mut().for_each(|d| {
+        d.update_dependent_datasource_captions(&captions);
+    });
+    Ok(datasources)
+}
+
+impl RawDatasource {
+    pub fn find_table(&self, col_name: &str) -> String {
+        if let Some(ref conn) = self.connection {
+            conn.metadata_records.columns.get(col_name)
+                .map(|meta| meta.table.clone())
+                .unwrap_or_default()
+        } else {
+            String::new()
+        }
+    }
+
+    pub fn get_table_aggregation(&self, table: &str) -> Option<String> {
+        self.connection.as_ref()
+            .and_then(|conn| conn.metadata_records.capabilities.get(table)
+                .cloned())
+    }
+
+    fn update_dependent_datasource_captions(&mut self, captions: &HashMap<String, String>) {
+        self.dependencies.values_mut().for_each(|dep| {
+            dep.caption = captions.get(&dep.name).cloned().unwrap_or_default();
+        });
+    }
+}
+
+/// returns the caption for the column surrounded by `[]`.
+fn get_column_captioned(c: &ColumnMeta) -> Cow<str> {
+    if c.caption.is_empty() {
+        Cow::from(c.name.as_str())
```

Comment:
We need a Cow here because we might be returning either an &str (reference to the name) or a String (bracketed caption). We can't return an &str since the bracketed caption was created within this method and we would need to transfer ownership of that String to the caller. 
We could return a String (owned) instead. However, that requires cloning the name reference, since we can't take ownership of `c.name` (as c is a reference to a ColumnMeta). Cloning would waste memory, especially in a commonly used method like this where we only really need a string reference.

Cow<str> provides us with the way to provide either an owned String or an &str and won't clone the &str unless the caller needs to. So, overall, this is a bit of a memory optimization for the code.